File: 객지/week5/week5_Lecture/src/main/kotlin/section2_java/Dog.java
package section2_java;

public class Dog { // ① 클래스의 정의
    // ② 프로퍼티들(속성)
    String type;
    String name;
    int age;

    // ③ 메서드들(함수)
    void bark() {
        System.out.println(name+"(이/가)  짖고 있습니다.");
    }
    void eat() {
        System.out.println(name+"(이/가) 식사 중입니다.");
    }

    public static void main(String[] args) {
        Dog aDog = new Dog();  // ④ 클래스의 생성자를 통한 객체의 생성
        aDog.type = "진도개";  // ⑤ 객체의 프로퍼티에 값 할당
        aDog.name = "화랑";
        aDog.age = 3;

        // ⑥ 객체의 멤버 프로퍼티 읽기
        System.out.println("aDog.type: "+aDog.type +
                ", aDog.name: "+ aDog.name +
                ", aDog.age: " + aDog.age);
        aDog.bark(); // ⑦ 객체의 멤버 메서드의 사용
        aDog.eat();
    }
}

File: 객지/week5/week5_Lecture/src/main/kotlin/section3/mainconstructor/initBlock/propertyAsParam/Dog.kt
package section3.mainconstructor.initBlock.propertyAsParam

class Dog(var type: String = "Dog", var name: String?, var age: Int = 0) {
    init {
        if (type!="삽살개" && type!="진도개" && type!="Dog") {
            println("${type} is an invalid type!!: Dog 로 초기화")
            type = "Dog"
        }
        this.name = name ?: "noName"
        if (age !in 0..30) {
            println("${age} is an invalid age!!: 0으로 초기화")
            age = 0
        }
    }

    fun bark() {
        println("$name(이/가)  짖고 있습니다.")
    }

    fun eat() {
        println("$name(이/가) 식사 중입니다.")
    }
}

fun main() {
    println("객체생성 요청: Dog(\"진도개\",\"화랑\",3)")
    val aDog = Dog("진도개", "화랑", 3)
    println("종류= ${aDog.type}")
    println("이름= ${aDog.name}")
    println("나이= ${aDog.age}")

    println("\n객체생성 요청: Dog(\"진도개2\",null,100)")
    val bDog = Dog("진도개2", null, 100)
    println("종류= ${bDog.type}")
    println("이름= ${bDog.name}")
    println("나이= ${bDog.age}")

    println("\n객체생성: Dog(name=\"해피\")")
    val cDog = Dog(name = "해피")
    println("종류= ${cDog.type}")
    println("이름= ${cDog.name}")
    println("나이= ${cDog.age}")
}